pilot/pkg/controllers/ipallocate/ipallocate.go

```diff
@@ -0,0 +1,402 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	klabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/api/meta/v1alpha1"
+	networkingv1alpha3 "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	autoallocate "istio.io/istio/pilot/pkg/serviceregistry/serviceentry"
+	kubelib "istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	istiolog "istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/slices"
+)
+
+var log = istiolog.RegisterScope("ip-autoallocate", "IP autoallocate controller")
+
+type IPAllocate struct {
+	serviceEntryClient kclient.Client[*networkingv1alpha3.ServiceEntry]
+	stopChan           <-chan struct{}
+	queue              controllers.Queue
+
+	// This controller is not safe for concurency but it exists outside the critical path performing important but minimal functionality in a single thread.
+	// If we want the multi-thread this controller we must add locking around accessing the allocators which would otherwise be racy
+	v4allocator *ipAllocator
+	v6allocator *ipAllocator
+}
+
+type conflictDetectedEvent struct {
+	conflictingResourceIdentifier types.NamespacedName
+	conflictingAddress            netip.Addr
+}
+
+const (
+	controllerName = "IP Autoallocate"
+	// these are the ranges of the v1 logic, do we need to choose new ranges?
+	IPV4Prefix = "240.240.0.0/16"
+	IPV6Prefix = "2001:2::/48"
+)
+
+func NewIPAllocate(stop <-chan struct{}, c kubelib.Client) *IPAllocate {
+	client := kclient.New[*networkingv1alpha3.ServiceEntry](c)
+	allocator := &IPAllocate{
+		serviceEntryClient: client,
+		stopChan:           stop,
+		// MustParsePrefix is OK because these are const. If we allow user configuration we must not use this function.
+		v4allocator: newIPAllocator(netip.MustParsePrefix(IPV4Prefix)),
+		v6allocator: newIPAllocator(netip.MustParsePrefix(IPV6Prefix)),
+	}
+	allocator.queue = controllers.NewQueue(controllerName, controllers.WithGenericReconciler(allocator.reconcile), controllers.WithMaxAttempts(5))
+	client.AddEventHandler(controllers.ObjectHandler(allocator.queue.AddObject))
+	return allocator
+}
+
+func (c *IPAllocate) Run(stop <-chan struct{}) {
+	log.Debugf("starting %s controller", controllerName)
+	kubelib.WaitForCacheSync(controllerName, stop, c.serviceEntryClient.HasSynced)
+	// it is important that we always warm cache before we try to run the queue
+	// failing to warm cache first could result in allocating IP which are already in use
+	c.warmCache()
+	// logically the v1 behavior is a state of the world function which cannot be replicated in reconcile
+	// if we wish to replicate v1 assignment behavior for existing serviceentry it should go here and must
+	// include a change to respect the used IPs from the warmed cache begin reconcile
+	c.queue.Run(stop)
+	c.serviceEntryClient.ShutdownHandlers()
+}
+
+// warmCache reads all serviceentries and records any auto-assigned addresses as in use
+func (c *IPAllocate) warmCache() {
```

Comment:
Maybe I'm confused - but doesn't 'wait for sync' guarantee that the cache will be warm ? Informer should get all the existing objects as callbacks.

Reply:
Warm cache is making sure our internal data structures are populated. It's not the informer cache we're worried about here, although we do want that to be ready before we try warming our own cache as well.

Reply:
Renamed to avoid any unintended association with the underlying client synching.